lib/auth/access.go

```diff
@@ -144,3 +148,71 @@ func (a *Server) DeleteLock(ctx context.Context, lockName string) error {
 	}
 	return nil
 }
+
+// checkRoleRulesConstraint checks if the request will result in having
+// no roles with rules to upsert roles.
+func (s *Server) checkRoleRulesConstraint(ctx context.Context, targetRole types.Role, request string) error {
+	rolesWithUpdateRolesRule, err := s.getRolesWithUpdateRolesRule(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	isRoleLoosingUpdateRolesRule := Contains(rolesWithUpdateRolesRule, targetRole.GetName()) && !roleHasUpdateRolesRule(targetRole)
+
+	if !isRoleLoosingUpdateRolesRule {
+		return nil
+	}
+
+	allUsers, err := s.Identity.GetUsers(false)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	localUsersWithPermissionToEditRoles := Filter(allUsers, func(u types.User) bool {
+		isLocalUser := u.GetCreatedBy().Connector == nil
+
+		rolesWithoutTarget := Filter(u.GetRoles(), func(role string) bool {
+			return role != targetRole.GetName()
+		})
+
+		return isLocalUser && Some(rolesWithoutTarget, func(role string) bool {
+			return Contains(rolesWithUpdateRolesRule, role)
+		})
+	})
+
+	if len(localUsersWithPermissionToEditRoles) > 0 {
```

Comment:
are we not going to support having a cluster with zero local users? was there a product decision on that?